### PR TITLE
ndc-spec 0.1.3 with filter/order by nested fields

### PR DESCRIPTION
**Pull request by soupi**
### What

1. Update ndc-spec to version [0.1.3](https://github.com/hasura/ndc-spec/releases/tag/v0.1.3).
2. Implemented the new capabilities: filter and order by nested fields.

#### Nested fields selection

We add SQL syntax for accessing an expression's field. The syntax is `(<expression>).<field>`

#### Order by nested fields

We add the field selection to the sort expression in the `ORDER BY` clause so it looks like this: `ORDER BY (column).field ASC`.
Implementation wise, we thread the field path like we do for the sort direction, and wrap the expressions in a NestedFieldSelect clause.

#### Filter by nested fields

- We directly wrap columns that contain field paths when they are used inside the predicate.
- We extract the type of a nested field by recursively looking it up in the environment.

### CODE REVIEW

File: crates/tests/databases-tests/src/postgres/snapshots/databases_tests__postgres__query_tests__predicates__filter_by_nested_field.snap [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+            {
+              "name": "Legolas",
+              "popularity": "200"
+            },
+            {
+              "name": "Gimli",
+              "popularity": "300"
+            }
+          ],
+          "name": "Fellowship"
```
Comment (soupi) [inline crates/tests/databases-tests/src/postgres/snapshots/databases_tests__postgres__query_tests__predicates__filter_by_nested_field.snap:24]: We filtered by this nested field.

File: crates/tests/databases-tests/src/postgres/snapshots/databases_tests__postgres__query_tests__sorting__order_by_nested_field.snap [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            {
+              "name": "Legolas",
+              "popularity": "200"
+            },
+            {
+              "name": "Gimli",
+              "popularity": "300"
+            }
+          ],
+          "name": "Fellowship"
```
Comment (soupi) [inline crates/tests/databases-tests/src/postgres/snapshots/databases_tests__postgres__query_tests__sorting__order_by_nested_field.snap:58]: we sorted by this field, descending.

File: crates/tests/tests-common/goldenfiles/select_array_composite_field.json [hunk 3]
```diff
@@ -12,7 +12,20 @@
...
+          "order_direction": "asc",
+          "target": {
+            "type": "column",
+            "name": "characters",
+            "path": []
+          }
+        }
+      ]
+    },
+    "limit": 1
```
Comment (soupi) [inline crates/tests/tests-common/goldenfiles/select_array_composite_field.json:28]: We added more data to the table, this was done to keep the previous output for the test.

File: crates/query-engine/translation/src/translation/error.rs [hunk 4]
```diff
@@ -48,11 +48,15 @@ pub enum Error {
 
 /// Capabilities we don't currently support.
 #[derive(Debug, Clone)]
-pub enum UnsupportedCapabilities {}
+pub enum UnsupportedCapabilities {
+    FieldArguments,
```
Comment (soupi) [inline crates/query-engine/translation/src/translation/error.rs:52]: We don't support field arguments atm.

File: crates/query-engine/translation/src/translation/helpers.rs [hunk 5]
```diff
@@ -554,3 +554,30 @@ impl NativeQueries {
...
+/// A newtype wrapper around an ndc-spec type which represents accessing a nested field.
+#[derive(Debug, Clone)]
+pub struct FieldPath(pub Vec<String>);
+
+impl From<&Option<Vec<String>>> for FieldPath {
+    fn from(field_path: &Option<Vec<String>>) -> Self {
+        FieldPath(match field_path {
+            // The option has no logical function other than to avoid breaking changes.
+            None => vec![],
+            Some(vec) => vec.clone(),
```
Comment (soupi) [inline crates/query-engine/translation/src/translation/helpers.rs:567]: Potentially we could use refcast, but this is probably overkill for what is probably a small amount of fairly short vectors.

File: crates/query-engine/translation/src/translation/query/sorting.rs [hunk 6]
```diff
@@ -291,10 +318,23 @@ fn translate_order_by_target_group(
 /// or a select query describing how to reach the columns.
 enum ColumnsOrSelect {
     /// Columns represents target columns that are referenced from the current table.
-    Columns(Vec<(usize, models::OrderDirection, sql::ast::ColumnReference)>),
-    /// Select represents a select query which contain the requested columns.
+    Columns(
```
Comment (plcplc) [inline crates/query-engine/translation/src/translation/query/sorting.rs:321]: Nice!

**Review by plcplc — APPROVED** [review]
LGTM.

Love the improved commentary.

**Pull request merged**